sdk/go/x/auto/local_workspace.go

```diff
@@ -0,0 +1,719 @@
+	stdout, stderr, errCode, err := l.runPulumiCmdSync(ctx, "plugin", "ls", "--json")
+	if err != nil {
+		return nil, errors.Wrap(newAutoError(err, stdout, stderr, errCode), "could not list list")
+	}
+	var plugins []workspace.PluginInfo
+	err = json.Unmarshal([]byte(stdout), &plugins)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to unmarshal plugin response")
+	}
+	return plugins, nil
+}
+
+// Program returns the program `pulumi.RunFunc` to be used for Preview/Update if any.
+// If none is specified, the stack will refer to Project Settings for this information.
+func (l *LocalWorkspace) Program() pulumi.RunFunc {
+	return l.program
+}
+
+// SetProgram sets the program associated with the Workspace to the specified `pulumi.RunFunc`
+func (l *LocalWorkspace) SetProgram(fn pulumi.RunFunc) {
+	l.program = fn
+}
+
+func (l *LocalWorkspace) runPulumiCmdSync(
+	ctx context.Context,
+	args ...string,
+) (string, string, int, error) {
+	var env []string
+	if l.PulumiHome() != nil {
+		homeEnv := fmt.Sprintf("%s=%s", pulumiHomeEnv, *l.PulumiHome())
+		env = append(env, homeEnv)
+	}
+	return runPulumiCommandSync(ctx, l.WorkDir(), env, args...)
+}
+
+// NewLocalWorkspace creates an configures a LocalWorkspace. LocalWorkspaceOptions can be used to
+// configure things like the working directory, the program to execute, and to seed the directory with source code
+// from a git repository.
+func NewLocalWorkspace(ctx context.Context, opts ...LocalWorkspaceOption) (Workspace, error) {
+	lwOpts := &localWorkspaceOptions{}
+	// for merging options, last specified value wins
+	for _, opt := range opts {
+		opt.applyLocalWorkspaceOption(lwOpts)
+	}
+
+	var workDir string
+
+	if lwOpts.WorkDir != "" {
+		workDir = lwOpts.WorkDir
+	} else {
+		dir, err := ioutil.TempDir("", "pulumi_auto")
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to create tmp directory for workspace")
+		}
+		workDir = dir
+	}
+
+	if lwOpts.Repo != nil {
+		// now do the git clone
+		projDir, err := setupGitRepo(ctx, workDir, lwOpts.Repo)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace, unable to enlist in git repo")
+		}
+		workDir = projDir
+	}
+
+	var program pulumi.RunFunc
+	if lwOpts.Program != nil {
+		program = lwOpts.Program
+	}
+
+	var pulumiHome *string
+	if lwOpts.PulumiHome != nil {
+		pulumiHome = lwOpts.PulumiHome
+	}
+
+	l := &LocalWorkspace{
+		workDir:    workDir,
+		program:    program,
+		pulumiHome: pulumiHome,
+	}
+
+	if lwOpts.Project != nil {
+		err := l.WriteProjectSettings(ctx, lwOpts.Project)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace, unable to save project settings")
+		}
+	}
+
+	for fqsn := range lwOpts.Stacks {
+		s := lwOpts.Stacks[fqsn]
+		err := l.WriteStackSettings(ctx, fqsn, &s)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create workspace")
+		}
+	}
+
+	return l, nil
+}
+
+type localWorkspaceOptions struct {
```

Comment:
I think that we should make this type an option in and of itself. This would allow easy manipulation of options after the fact.

Reply:
Not sure I grok. Would you mind explaining more or perhaps provide an example?